drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -552,100 +552,145 @@ virtio_vdpa_vring_state_set(int vid, int vq_idx, int state)
-		DRV_LOG(ERR, "%s failed to get VM memory layout ret:%d",
-					priv->vdev->device->name, ret);
-		goto exit;
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
 	}
 
-	vfio_container_fd = priv->vfio_container_fd;
+	mem = priv->mem;
+	if (mem == NULL) {
+		DRV_LOG(INFO, "No mem is registered: %s", vdev->device->name);
+		return 0;
+	}
 
 	for (i = 0; i < mem->nregions; i++) {
 		struct rte_vhost_mem_region *reg;
 
 		reg = &mem->regions[i];
 		DRV_LOG(INFO, "%s, region %u: HVA 0x%" PRIx64 ", "
 			"GPA 0x%" PRIx64 ", size 0x%" PRIx64 ".",
-			"DMA unmap", i,
+			"DMA clean up unmap", i,
 			reg->host_user_addr, reg->guest_phys_addr, reg->size);
 
-		ret = rte_vfio_container_dma_unmap(vfio_container_fd,
+		ret = rte_vfio_container_dma_unmap(priv->vfio_container_fd,
 			reg->host_user_addr, reg->guest_phys_addr,
 			reg->size);
 		if (ret < 0) {
-			DRV_LOG(ERR, "%s DMA unmap failed ret:%d",
+			DRV_LOG(ERR, "%s DMA clean up failed ret:%d",
 						priv->vdev->device->name, ret);
-			goto exit;
+			free(mem);
+			priv->mem = NULL;
+			return ret;
 		}
 	}
 
-exit:
 	free(mem);
-	return ret;
+	priv->mem = NULL;
+	return 0;
 }
+static inline bool
+virtio_vdpa_find_mem_reg(struct rte_vhost_mem_region *key,struct rte_vhost_memory *mem)
+{
+	uint32_t i = 0;
+	struct rte_vhost_mem_region *reg;
+
+	if (mem == NULL)
+		return false;
 
+	for (i = 0; i < mem->nregions; i++) {
+		reg = &mem->regions[i];
+		if ((reg->host_user_addr == key->host_user_addr) &&
+			(reg->guest_phys_addr == key->guest_phys_addr) &&
+			(reg->size == key->size))
+			return true;
+	}
+	return false;
+}
 static int
-virtio_vdpa_dma_map(struct virtio_vdpa_priv *priv)
+virtio_vdpa_dev_set_mem_table(struct rte_vdpa_device *vdev)
 {
-	uint32_t i = 0, j;
+	uint32_t i = 0;
 	int ret;
-	struct rte_vhost_memory *mem = NULL;
+	struct rte_vhost_memory *cur_mem = NULL;
 	struct rte_vhost_mem_region *reg;
-	int vfio_container_fd = priv->vfio_container_fd;
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
 
-	ret = rte_vhost_get_mem_table(priv->vid, &mem);
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+
+	ret = rte_vhost_get_mem_table(priv->vid, &cur_mem);
 	if (ret < 0) {
 		DRV_LOG(ERR, "%s failed to get VM memory layout ret:%d",
 					priv->vdev->device->name, ret);
-		goto exit;
+		return ret;
 	}
 
-	for (i = 0; i < mem->nregions; i++) {
-		reg = &mem->regions[i];
-		DRV_LOG(INFO, "%s, region %u: HVA 0x%" PRIx64 ", "
-			"GPA 0x%" PRIx64 ", size 0x%" PRIx64 ".",
-			"DMA map", i,
-			reg->host_user_addr, reg->guest_phys_addr, reg->size);
-
-		ret = rte_vfio_container_dma_map(vfio_container_fd,
-			reg->host_user_addr, reg->guest_phys_addr,
-			reg->size);
-		if (ret < 0) {
-			DRV_LOG(ERR, "%s DMA map failed ret:%d",
-						priv->vdev->device->name, ret);
-			goto exit;
+	/* Unmap reagion dind't exsit in current */
+	if (priv->mem) {
```

Comment:
better to have two different helper routines for map and unmap and call them here.